File: Strings/validParentheses.js
// Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

// An input string is valid if:

// Open brackets must be closed by the same type of brackets.
// Open brackets must be closed in the correct order.


// Example 1:

// Input: s = "()"
// Output: true
// Example 2:

// Input: s = "()[]{}"
// Output: true
// Example 3:

// Input: s = "(]"
// Output: false

const isValid = s => {
  let stack = []
  let paren = s.split('')

  for (let char of s) {
    if (char === '{') stack.push('}')
    else if (char === '(') stack.push(')')
    else if (char === '[') stack.push(']')
    else if (stack.length === 0 || char !== stack.pop()) return false
  }

  if (stack.length === 0) return true
  return false
}

File: Arrays/valid_parentheses.py
# Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

# An input string is valid if:

# Open brackets must be closed by the same type of brackets.
# Open brackets must be closed in the correct order.
# Every close bracket has a corresponding open bracket of the same type.


def isValid(s: str) -> bool:
    close_to_open = {'}':'{', ']':'[', ')':'('}
    stack = []

    for c in s:
        if c in close_to_open:
            if stack and stack[-1] == close_to_open[c]:
                stack.pop()
            else:
                return False
        else:
            # we got an open paren so we can add to stack
            stack.append(c)
    return True if not stack else False